Dockerfile

```diff
@@ -1,35 +1,60 @@
-# build stage
-FROM golang:1-alpine AS build-env
-RUN apk add --no-cache \
-    build-base \
-    linux-headers \
-    coreutils \
-    binutils \
-    libtool \
-    musl-dev \
-    git \
-    make \
-    gcc \
-    libc-dev \
-    zip \
-    ca-certificates \
-    libcap
-
-ENV GO111MODULE=on \
-    CGO_ENABLED=0
-    
-WORKDIR /src
+# get newest certificates
+FROM --platform=$BUILDPLATFORM alpine:3.16 AS ca-certs
+RUN apk add --no-cache ca-certificates
+RUN --mount=type=cache,target=/etc/ssl/certs \
+    update-ca-certificates 2>/dev/null || true
 
+# zig compiler
+FROM --platform=$BUILDPLATFORM ghcr.io/euantorano/zig:master AS zig-env
+
+# build environment
+FROM --platform=$BUILDPLATFORM golang:1-alpine AS build
+
+# required arguments(buildx will set target)
+ARG VERSION
+ARG BUILD_TIME
+ARG TARGETOS
+ARG TARGETARCH
+ARG TARGETVARIANT
+
+# set working directory
+WORKDIR /go/src
+
+# download packages
 COPY go.mod go.sum ./
-RUN go mod download
+RUN --mount=type=cache,target=/go/pkg \
+    go mod download
 
 # add source
-ADD . .
+COPY . .
 
-ARG opts
-RUN env ${opts} make build-static && \
-    chown 100 /src/bin/blocky && \
-    setcap 'cap_net_bind_service=+ep' /src/bin/blocky
+# setup go+zig
+COPY --from=zig-env /usr/local/bin/zig /usr/local/bin/zig
+ENV PATH="/usr/local/bin/zig:${PATH}"
+RUN --mount=type=cache,target=/go/pkg \
+    go install github.com/dosgo/zigtool/zigcc@latest && \
+    go install github.com/dosgo/zigtool/zigcpp@latest && \
+    go env -w GOARM=${TARGETVARIANT##*v}
+ENV CC="zigcc" \
+    CXX="zigcpp" \
+    CGO_ENABLED=0 \
+    GOOS="linux" \
+    GOARCH=$TARGETARCH
+
+#add libcap
+RUN apk add --no-cache libcap
+
+# build binary 
+RUN --mount=type=bind,target=. \
+    --mount=type=cache,target=/root/.cache/go-build \ 
+    --mount=type=cache,target=/go/pkg \
+    go build \
+    -tags static \
+    -v \
+    -ldflags="-X github.com/0xERR0R/blocky/util.Version=${VERSION} -X github.com/0xERR0R/blocky/util.BuildTime=${BUILD_TIME} -X github.com/0xERR0R/blocky/util.Architecture=${TARGETARCH}${TARGETVARIANT}" \
+    -o /bin/blocky && \
+    chown 100 /bin/blocky && \
+    setcap 'cap_net_bind_service=+ep' /bin/blocky
```

Comment:
Is there a reason this was moved out of the `Makefile`?  
My first reaction is that it would make sense to keep the `util.*` handling in the same spot.

Reply:
Build time reduction.
build-static was only used inside the Docker file and didn't work for cross compilation.
Currently all necessary tools for it are defined in it. This ensures that the image can be build for multiple target systems without rerunning the environment setup again.
Moving these steps to the make file would create overhead without a benefit in my opinion.

Reply:
The main benefit is making the `-X` linker flags DRY but I get your point about user exposure.  
What do you think about having the `Makefile` support passing extra build flags? That way the `Dockerfile` could call `make build EXTRA_GO_BUILD_ARGS="-static [...]"` and we don't duplicate `-X` anywhere.

Reply:
I get your point and will try to change it.